runtime/ms-rest-azure/lib/filters/rpRegistrationFilter.js

```diff
@@ -0,0 +1,121 @@
+        if (rpName) {
+          urlPrefix = exports.extractSubscriptionUrl(options.url);
+          return exports.registerRP(urlPrefix, rpName, options, (err, result) => {
+            if (err) {
+              return callback(err);
+            }
+            if (result) {
+              //retry the original request
+              options.headers['x-ms-client-request-id'] = utils.generateUuid();
+              return request(options.url, options, (error, response, body) => {
+                return callback(error, response, response.body);
+              });
+            } else {
+              return callback(err, response, body);
+            }
+          });
+        }
+      }
+      return callback(err, response, body);
+    });
+  };
+};
+
+exports.checkRPNotRegisteredError = function checkRPNotRegisteredError(body) {
+  let result, responseBody;
+  if (body) {
+    try {
+      responseBody = JSON.parse(body);
+    } catch (err) {
+      //do nothing;
+    }
+    if (responseBody.error && responseBody.error.message && 
+    responseBody.error.code && responseBody.error.code === 'MissingSubscriptionRegistration') {
+      let matchRes = responseBody.error.message.match(/'(\w+\.?\w+)'*/ig);
+      if (matchRes && matchRes[1]) {
+        result = matchRes[1].slice(1, matchRes[1].length-1);
+      }
+    }
+  }
+  return result;
+};
+
+/**
+ * Extract the first part of the URL, just after subscription: 
+ * https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/
+ */
+exports.extractSubscriptionUrl = function extractSubscriptionUrl(url) {
+  let result;
+  let matchRes = url.match(/.*\/subscriptions\/[a-f0-9-]+\//ig);
+  if (matchRes && matchRes[0]) { 
+    result = matchRes[0];
+  } else {
+    throw new Error(`Unable to extract subscriptionId from the given url - ${url}.`);
+  }
+  return result;
+}
+
+exports.registerRP = function regiserRP(urlPrefix, provider, originalRequest, callback) {
+  let postUrl = `${urlPrefix}providers/${provider}/register?api-version=2016-02-01`;
+  let getUrl = `${urlPrefix}providers/${provider}?api-version=2016-02-01`;
+  let reqOptions = {
+    headers: originalRequest.headers
+  };
+  reqOptions.headers['x-ms-client-request-id'] = utils.generateUuid();
+  return request.post(postUrl, reqOptions, (err, response, body) => {
+    if (response.statusCode !== 200) {
+      return callback(new Error(`Autoregistration of ${provider} failed. Please try registering manually.`));
+    }
+    return exports.getRegistrationStatus(getUrl, 0, originalRequest, (err, result) => {
+      if (err) {
+        return callback(err, result);
+      }
+      return callback(null, result);
+    });
+  });
+};
+
+exports.getRegistrationStatus = function getRegistrationStatus(url, count, originalRequest, callback) {
+  let reqOptions = {
+    headers: originalRequest.headers
+  };
+  reqOptions.headers['x-ms-client-request-id'] = utils.generateUuid();

```

Comment:
done. Created a generic function to do this and documented the reason for doing this.